contracts/ring.md

```diff
@@ -0,0 +1,340 @@
+worked out).
+- A contract may define a type state encapsulating its local
+  state. The state must be initialised when instantiating a
+  contract. This is done by the init function which can take arbitrary
+  arguments and is called on contract instance creation.
+- Contracts have a public API, comprising the functions and types
+  annotated with the public keyword. These can be used from outside
+  the contract. Functions and types with no annotation (or internal)
+  are part of the internal API and can be used by contracts inheriting
+  (see below) the given contract. Functions and types annotated with
+  private can only be used locally.
+- Contracts can inherit one (or more?) other contract(s). In this case
+  the public functions (and types) of the inherited contract are
+  included in the public API and internal functions are included in
+  the internal API of the current contract. The state of the contract
+  contains the states of all inherited contracts as well as the local
+  state. However, the state of an inherited contract cannot be
+  accessed other than through internal functions defined by that
+  contract.
+- Open question: should we allow overriding defined functions? I would
+  suggest no, but there might be compelling use cases that I'm
+  missing.
+
+### Mutable state
+
+Ring does not have arbitrary mutable state, but only a limited form of
+state associated with each contract instance.
+
+- Each contract defines a state type encapsulating its mutable state.
+- The value of the state is accessible from inside the contract
+  through an implicitly bound variable state.
+- State updates are performed by calling a function put : state =>
+  unit (possibly with a better name).
+- Aside from the put function (and similar functions for transactions
+  and events), the contract language is purely functional.
+- Functions modifying the state need to be annotated with the stateful keyword.
+
+To make it convenient to update parts of a deeply nested state Ring
+provides special special syntax for map/record updates.  Open
+question: we likely want to make it possible to have immutable state
+(parameters). Keep separate from mutable state or annotate certain
+fields as immutable?
+
+### Types
+Ring has the following types:
+
+| Type    | Description                     | Example
+| ------- | ------------------------------- | -------:
+| uint    | A 256 bit integer               | ```42```
+| int     | A 256 bit 2-complement integer  | ```-1```
+| address | A 256 bit number given as a hex | ```ff00```
+| bool    | A Boolean                       | ```true```
+| string  | An array of bytes               | ```“Foo”```
+| list    | A homogeneous immutable singly linked list. | ```[1, 2, 3]```
+| tuple   | An ordered heterogeneous array   | ```(42, “Foo”, true)```
+| record  | An immutable key value store with fixed key names and typed values | ``` type balance = { owner: address, value: uint } ```
```

Comment:
Is order of keys in definition relevant?

Reply:
No. But in the implementation/internal representation it has to be defined, and I think sorted on key is the way to go.